src/electionguard/decrypt_with_shares.py

```diff
@@ -0,0 +1,213 @@
+from typing import Dict, Optional, Tuple, Union
+
+from .ballot import CiphertextAcceptedBallot, CiphertextBallotSelection
+from .decryption_share import (
+    DecryptionShare,
+    CiphertextDecryptionSelection,
+    get_spoiled_shares_for_selection,
+    get_tally_shares_for_selection,
+)
+from .dlog import discrete_log
+from .election import CiphertextElectionContext
+from .group import ElementModP, ElementModQ, mult_p, div_p
+from .tally import (
+    CiphertextTally,
+    PlaintextTally,
+    CiphertextTallyContest,
+    PlaintextTallyContest,
+    CiphertextTallySelection,
+    PlaintextTallySelection,
+)
+from .logs import log_warning
+from .types import BALLOT_ID, CONTEST_ID, GUARDIAN_ID, SELECTION_ID
+
+AVAILABLE_GUARDIAN_ID = GUARDIAN_ID
+MISSING_GUARDIAN_ID = GUARDIAN_ID
+
+GUARDIAN_PUBLIC_KEY = ElementModP
+
+ELECTION_PUBLIC_KEY = ElementModP
+
+CiphertextSelection = Union[CiphertextBallotSelection, CiphertextTallySelection]
+
+# The methods in this file can be used to decrypt values if private keys or nonces are not known
+# and the key ceremony is used to share secrets among a quorum of guardians
+
+
+def decrypt_selection_with_decryption_shares(
+    selection: CiphertextSelection,
+    shares: Dict[
+        GUARDIAN_ID, Tuple[ELECTION_PUBLIC_KEY, CiphertextDecryptionSelection]
+    ],
+    extended_base_hash: ElementModQ,
+    suppress_validity_check: bool = False,
+) -> Optional[PlaintextTallySelection]:
+    """
+    Decrypt the specified `CiphertextTallySelection` with the collection of `ElementModP` decryption shares
+
+    :param selection: a `CiphertextTallySelection`
+    :param shares: the collection of shares to decrypt the selection
+    :param extended_base_hash: the extended base hash code (𝑄') for the election
+    :return: a `PlaintextTallySelection` or `None` if there is an error
+    :param suppress_validity_check: do not validate the encryption prior to decrypting (useful for tests)
+    """
+
+    if not suppress_validity_check:
+        # Verify that all of the shares are computed correctly
+        for guardian_id, share in shares.items():
+            public_key, decryption = share
+            # verify we have a proof or recovered parts
+            if decryption.proof is None and decryption.recovered_parts is None:
+                log_warning(
+                    f"decrypt selection failed for guardian: {guardian_id} selection: {selection.object_id} with missing data"
+                )
+                return None
+
+            if decryption.proof is not None and decryption.recovered_parts is not None:
+                log_warning(
```

Comment:
I don't particularly love these log messages. I wish we had something like:
`log_decryption_fail()` that took or something similar that was a little less verbose to keep the lines cleaner. It feels like it should be error handling. The more I see this the more it feels a bit like code smell.